### ISSUE TITLE
deterministic state ids

### ISSUE THREAD

**Issue by krancour** [akuity/kargo#299]
Currently, state (bill of materials) IDs are system-assigned UUIDs.

There are a few implications to this. One is that for "zero environments" that subscribe directly to upstream repositories, each time the repos are checked for new materials, the resulting state is akways "new" in that it gets a fresh new UUID. To determine if it's _actually_ new, we need to do a comparison of its _materials_ to the top state in the stack of previously known available states. This, on its own isn't terrible...

The second problem is an edge case where a zero env has been deleted/replaced. The new one would scan upstream repos for materials and find the same materials as before, assign a new ID as usual, but would fail to determine it's a duplicate of a state that may already be known to downstream environments.

Basically, this is a way for duplicate states with different IDs to enter the system.

If state IDs were deterministically generated as a function of a state's materials, none of the second issue wouldn't be an issue and the solution to the first issue would be simpler.

What I propose is simple: Replace UUIDs with a hash of a canonicalized representation of the state's materials. For example: 

`hash(<git repo 0>|<git repo 1>|...|<git repo n>|<image 0>|<image 1>|...|<image n>|<chart 0>|<chart 1>|...|<chart n>)`
